### PR TITLE
Edit comment QoL tweaks

### PR DESCRIPTION
**Pull request by stwalkerster**
* Show request status on edit comment
* Show a warning when editing a flagged comment on a non-closed request
* Show link to edit comment and the comment in-situ on the request from the log page.
* Fix issue where it was impossible to only unflag a comment via the edit comment screen
* Edit comment screen now returns you to either the view request page or the flagged comments page, depending on which you used to edit the comment.
* Hides flagged comments from view when request is unreserved (#777)
* Handles some extra edge cases for when a user has access to the edit interfaces, but doesn't have `RequestData::alwaysSeePrivateData`
* Adds an IRC ping for outstanding flagged comments (#777)

### CODE REVIEW

File: includes/Pages/PageEditComment.php
Comment (dqwiki) [inline includes/Pages/PageEditComment.php:161]: The security of this function should be checked from the inside and not allow implementation to possibly breach security.

**Review by dqwiki — CHANGES_REQUESTED** [review]
Internal function security is a standard we should set forth on so that anything outside of protected security functions does not have a play in actual security. Maybe this requires some form of bigger audit of all our functions, but the less replicating of security functions, the more secure they will be.

File: includes/Pages/PageListFlaggedComments.php
Comment (stwalkerster) [reply includes/Pages/PageListFlaggedComments.php:51]: Perhaps I'm not understanding your point here.

The `RequestData::alwaysSeePrivateData` permission already exists, and is already in widespread use in Production today.

For wider context, private data is always hidden on the tool unless (at least) one of three situations is true:
* The user has reserved the request the private data refers to
* The user has browsed to the hash link for the request, having received this from whoever reserved the request
* The user has the `RequestData::alwaysSeePrivateData` permission via one of the roles granted that permission - currently `toolRoot`, `admin`, or `checkuser`

Are you suggesting that this third point should be removed entirely (noting that it would remove the exemption from CUs and admins and force members of those groups to reserve each request before seeing the data)?

This patch more widely hides the content of flagged comments from those who are not able to see private data, per your request in https://github.com/enwikipedia-acc/waca/issues/777#issuecomment-1360758330. This effectively treats flagged comments as private data.

Historically, ListFlaggedComments has had no private data on it by design, so it's not needed any of the private data protection that pages like ViewRequest have required. Thus at the moment (before this patch), the `ListFlaggedComments` page bypasses this protection entirely.

While we could make the (currently correct) assumption that everyone with the ability to browse to that page (by virtue of having the permission `PageListFlaggedComments::main`) also holds `RequestData::alwaysSeePrivateData`, and thus continue to skip the checks, I really don't want to rely on assumptions and thus I'm explicitly calling out and have written code for the case where that assumption isn't correct.

I think that not granting access to this page alongside alwaysSeePrivateData is a stupid configuration (hence my comment) as it forces the page to degrade to effectively a list of requests which happen to have flagged comments, rather than listing the comments in-line for quicker review.

File: includes/Pages/PageEditComment.php
Comment (stwalkerster) [reply includes/Pages/PageEditComment.php:161]: I'm really not understanding what you're suggesting I change here.

This function should indeed be `private` rather than `protected` - I'll fix that later - but your comment seems to suggest something more substantial?

**Comment by stwalkerster** [comment]
I've made some tweaks to this, namely changing some protected functions which were only used within the class itself into private methods. Arguably, some of these classes should be sealed/final too, but that's a wider change we ought to make across the tool.

I've also refactored the code around visibility checking within `ListFlaggedComments` to make it clearer on the decisions that are being made about what is/isn't visible.

Regarding this:
> Internal function security is a standard we should set forth on so that anything outside of protected security functions does not have a play in actual security.

I **strongly** disagree. Anything that interfaces with a DataObject should do a level of access checks where the data path is user-facing[^2]. Such user-facing security should be done at the action entrypoint level[^1]. An individual function should have a single, well-defined purpose. The code should generally be written in such a way that those security checks are clear and obvious (which admittedly isn't always the case).

For the case of ListFlaggedComments, the `main()` method can be fairly clearly seen to do a database retrieval, followed by loading the user and the relevant rights, then stepping through every loaded comment and applying security checks as appropriate.

It's arguable whether we should be splitting some of this logic out into separate [SOLID](https://enwp.org/SOLID) classes leaving the `Page*` classes in a purely coordination role. I still feel there's a good balance to be had between breaking the single-responsibility principle and ending up in full Enterprise Java land of 30+ classes like `RequestCommentTextSecurityValidationHelperCallbackFacade` doing almost nothing.


[^1]: By action entrypoints, I mean any `protected` method in any class referenced from [`RequestRouter`](https://github.com/enwikipedia-acc/waca/blob/5f75b59c628b392c8320d66e7b61d5ba4ae95b27/includes/Router/RequestRouter.php#L173) (and variants). To take PageBan as an example, that means the functions listed as actions (`set`, `remove`, and `show`), along with `main`.
[^2]: Things like maintenance scripts should be completely exempt from access controls; I don't want us to get anywhere close to a situation that a maintenance script fails to run because it doesn't quite have the necessary access to do it's job.

**Review by FastLizard4 — APPROVED** [review]
Code looks good and I don't see anything here that looks like it violates best practices for security or otherwise.

Regarding @dqwiki's comment:
> The security of this function should be checked from the inside and not allow implementation to possibly breach security.

I agree with all of Simon's points above related to this comment, but I also again just generally don't understand what is meant by this comment. Again, aside from a nitpick over `protected` vs. `private`, I don't see anything here that violates best practices.

Other comments in-line.

**Comment by FastLizard4** [comment]
For the record: I have no objections to dismissing @dqwiki's review at this point to unblock the PR, which I consider stale for the following reasons:

1. I genuinely cannot find the concerns discussed in her comments
2. Her comments are over a month old at this point, with no follow-up

Though I'd definitely love to hear more, at this point I don't see any reason to block the PR further than my already very delayed review has.

File: includes/Helpers/LogHelper.php
Comment (stwalkerster) [reply includes/Helpers/LogHelper.php:445]: Possibly a dedicated Smarty call for this, which would split it up into a template file and I think 6 lines of code?

And probably apply the same change for all things in this method, leading to about 10 more templates and ~50 more lines of code?

I'm really not sure if that's actually an improvement or not!

**Comment by dqwiki** [comment]
Sorry for delaying this for the length that I have.

My original comment wasn't to suggest that the security implementation itself is the issue, but simply the concern is the frequency in which we are reusing security based code (i.e. if someone can see private data). Replication of security based code is a security risk in and of itself, and my worry at the time was that we were stemming security based decisions (again, seeing if someone could see private data) into the code.

I have just reviewed now, and I don't know if I was looking at the wrong thing, but the current code doesn't present any of my concern. Weather I was too tired at the time of my comment to realize or if your changes in the meantime addressed the concern, I'm not sure.

It does raise the point though at some point in the future(tm) we should code security audit things and see if we detect any avoidable replication.

**Review by dqwiki — APPROVED** [review]

**Pull request merged**
